### PR TITLE
[DEVHAS-55] Component Detection Query Implementation

### PR DESCRIPTION
**Pull request by maysunfaisal**
Signed-off-by: Maysun J Faisal <maysunaneek@gmail.com>

FIxes https://issues.redhat.com/browse/DEVHAS-55

### Summary
- implements the Component Detection Query as per [CR doc](https://docs.google.com/document/d/1dfkTky3qFv6wTXdMX9Ag-bYFEt3OfBgRZOceYNSwyEU/edit?pli=1#heading=h.f1cye46tdm8s) & [API doc](https://docs.google.com/document/d/1CBeRA8LrdsDBG4AqFYqn4aovcd969lP_1-41OUAN2Bk/edit?pli=1#heading=h.f1cye46tdm8s)
- if `isMultiComponent` is false, only the root dir will be searched for a devfile; if it's true, the level 1 dir will be searched for a devfile
- if `isMultiComponent` is true and a `devfileURL` is mentioned it will err out
- the `componentDetected` field in `ComponentDetectionQuery` CR status will provide a mapping of all the devfiles detected and their detailed info pre-populated from the devfile into a `componentStub`; which can be used to create a `Component` CR
- The pre-populated data from the devfile is only fetched from the first devfile container component
- Tests
- Github action to lint the files

### Try it out
`kubectl apply -f config/samples/componentdetectionquery/componentdetectionquery-basic.yaml`
or

with a multi component repo https://github.com/maysunfaisal/multi-components (dummy application, not a working sample) and `isMultiComponent: true`

### CODE REVIEW

File: controllers/componentdetectionquery_controller.go [hunk 1]
```diff
@@ -47,9 +56,101 @@ type ComponentDetectionQueryReconciler struct {
...
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// If there is no component list in the map, the CR was just created
+	if len(componentDetectionQuery.Status.ComponentDetected) == 0 {
```
Comment (johnmcollier) [inline controllers/componentdetectionquery_controller.go:77]: Are there any potential scenarios where len(ComponentDetected) is 0, but we wouldn't want to go through this loop again? 

Like, say we exit out to do to some error.

**Review by johnmcollier — CHANGES_REQUESTED** [review]
Leaving a first pass of review comments

File: controllers/componentdetectionquery_controller_conditions.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	appstudiov1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+)
+
+func (r *ComponentDetectionQueryReconciler) SetCreateConditionAndUpdateCR(ctx context.Context, componentDetectionQuery *appstudiov1alpha1.ComponentDetectionQuery, createError error) {
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/componentdetectionquery_controller_conditions.go:29]: So I am wondering if in the future we can update it?

For instance, someone goes thru the import flow and provides a repo, HAC submits a query detection CR and sees there is no devfile. User then updates the repo; would HAC create a new CR? Probably yes at this point.

But that would also mean, HAC would need to delete these CRs? Otherwise there would be like old CRs lying around 🤔

File: controllers/componentdetectionquery_controller.go (hunk 1 above)
Comment (maysunfaisal) [reply controllers/componentdetectionquery_controller.go:77]: hmm interesting, yeah good point

wouldnt that fall under update in that case if we want to resubmit CR? I Wonder if its even possible. Consider this scenario:

- Submit a query CR
- No devfile, CR errors out
- user updates repo outside in git
- CR to recheck but no data is going to change, so will the server even take the request?

File: controllers/componentdetectionquery_controller.go [hunk 3]
```diff
@@ -47,9 +56,101 @@ type ComponentDetectionQueryReconciler struct {
...
+
+				devfilesMap, err = util.ReadDevfilesFromRepo(clonePath, 1)
+				if err != nil {
+					log.Error(err, fmt.Sprintf("Unable to find devfile(s) in repo %s, exiting reconcile loop %v", source.URL, req.NamespacedName))
+					r.SetCreateConditionAndUpdateCR(ctx, &componentDetectionQuery, err)
+					return ctrl.Result{}, err
+				}
+			} else {
+				log.Info(fmt.Sprintf("Since this is not a multi-component, attempt will be made to read devfile at the root dir... %v", req.NamespacedName))
+				rawURL, err := util.ConvertGitHubURL(source.URL)
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/componentdetectionquery_controller.go:108]: We discussed this and will use the scm file retriever package in the later sprints

File: config/samples/componentdetectionquery/componentdetectionquery-basic.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: ComponentDetectionQuery
+metadata:
+  name: componentdetectionquery-sample
+spec:
+  git:
+    # url: https://github.com/devfile-samples/devfile-sample-java-springboot-basic
+    # url: https://github.com/maysunfaisal/hiddendevfile
+    # url: https://github.com/maysunfaisal/hiddendirdevfile
+    url: https://github.com/maysunfaisal/hiddendirhiddendevfile
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply config/samples/componentdetectionquery/componentdetectionquery-basic.yaml]: updated

File: controllers/component_controller.go [hunk 5]
```diff
@@ -36,7 +36,8 @@ import (
 )
 
 const (
-	devfileName = "devfile.yaml"
+	devfileName       = "devfile.yaml"
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/component_controller.go]: updated

File: controllers/componentdetectionquery_controller_conditions.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/componentdetectionquery_controller_conditions.go]: updated to `Completed`

File: controllers/componentdetectionquery_controller.go [hunk 6]
```diff
@@ -18,21 +18,30 @@ package controllers
...
 
 // ComponentDetectionQueryReconciler reconciles a ComponentDetectionQuery object
 type ComponentDetectionQueryReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
+	Log    logr.Logger
 }
 
+const (
+	clonePathPrefix = "/tmp/appstudio/has"
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply controllers/componentdetectionquery_controller.go:42]: checked and confirmed to be fine

```
bash-4.4$ 
bash-4.4$ whoami
1000660000
bash-4.4$ id
uid=1000660000(1000660000) gid=0(root) groups=0(root),1000660000
```

```
bash-4.4$ pwd
/tmp
bash-4.4$ ls -la
total 8
drwxrwxrwt. 1 root       root  23 Jan 17 21:22 .
dr-xr-xr-x. 1 root       root  17 Jan 17 21:21 ..
drwxr-xr-x. 3 1000660000 root  17 Jan 17 21:22 appstudio
-rwx------. 1 root       root 701 Sep 14 16:20 ks-script-vzv5jj6c
-rwx------. 1 root       root 291 Sep 14 16:20 ks-script-y5xeen0d
bash-4.4$ 
bash-4.4$ cd appstudio/has/application-service-system/
bash-4.4$ 
bash-4.4$ 
bash-4.4$ pwd
/tmp/appstudio/has/application-service-system
bash-4.4$ ls -la
total 0
drwxr-xr-x. 3 1000660000 root  44 Jan 17 21:22 .
drwxr-xr-x. 3 1000660000 root  40 Jan 17 21:22 ..
drwxr-xr-x. 5 1000660000 root 134 Jan 17 21:22 componentdetectionquery-sample
```

File: pkg/util/util.go [hunk 7]
```diff
@@ -72,6 +87,24 @@ func ConvertGitHubURL(URL string) (string, error) {
 	return URL, nil
 }
 
+// CurlForDevfiles curls for various possible devfile locations in dir and returns the contents
+func CurlForDevfiles(dir string) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply pkg/util/util.go]: updated

File: pkg/util/util.go [hunk 8]
```diff
@@ -72,6 +87,24 @@ func ConvertGitHubURL(URL string) (string, error) {
...
+	for _, path := range validDevfileLocations {
+		devfilePath := dir + "/" + path
+		devfileBytes, err = CurlEndpoint(devfilePath)
+		if err == nil {
+			// if we get a 200, return
+			return devfileBytes, err
+		}
+	}
+
+	return nil, fmt.Errorf("unable to curl to any of known devfile locations in dir %s", dir)
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply pkg/util/util.go]: updated

File: pkg/util/util.go [hunk 9]
```diff
@@ -91,3 +124,95 @@ func CurlEndpoint(endpoint string) ([]byte, error) {
 
 	return nil, fmt.Errorf("received a non-200 status when curling %s", endpoint)
 }
+
+// CloneRepo clones the repoURL to clonePath
+func CloneRepo(clonePath, repoURL string) error {
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply pkg/util/util.go:121]: done

File: pkg/util/util.go [hunk 10]
```diff
@@ -91,3 +124,95 @@ func CurlEndpoint(endpoint string) ([]byte, error) {
...
+
+	return nil
+}
+
+// ReadDevfilesFromRepo attempts to read and return devfiles from the local path upto the specified depth
+func ReadDevfilesFromRepo(localpath string, depth int) (map[string][]byte, error) {
+	return searchDevfiles(localpath, 0, depth)
+}
+
+func searchDevfiles(localpath string, currentLevel, depth int) (map[string][]byte, error) {
```
[the remark this answers is not in the document]
Comment (maysunfaisal) [reply pkg/util/util.go:148]: done, for wrapper which calls this func too

File: controllers/componentdetectionquery_controller.go (hunk 1 above)
Comment (johnmcollier) [reply controllers/componentdetectionquery_controller.go:77]: Hmm, good point actually.

File: .github/workflows/pr.yml [hunk 11]
```diff
@@ -83,6 +83,10 @@ jobs:
             git --no-pager diff
             exit 1
           fi
+      - uses: dominikh/staticcheck-action@v1.1.0
```
Comment (johnmcollier) [inline .github/workflows/pr.yml:86]: Not something that needs to be addressed in this PR, but wondering: would it be better if this was its own PR check?

Comment (johnmcollier) [reply .github/workflows/pr.yml:86]: TBH, this could apply to other things in this PR check workflow as well

**Review by johnmcollier — APPROVED** [review]

File: .github/workflows/pr.yml (hunk 11 above)
Comment (maysunfaisal) [reply .github/workflows/pr.yml:86]: yeah we could clean up the workflow a bit
